### ISSUE TITLE
Animate.autoPlay

### ISSUE THREAD

**Issue by gskinner** [gskinner/flutter_animate#5]
Currently, to create an animation that doesn't play automatically, you need to use `onInit`:

``` dart
foo.animate(onInit: (controller) => controller.stop()).fade()
```

This works, but less semantically clear than having an `autoPlay` param.

However, implementation of such a feature has some ambiguities / concerns:

1. Should `Animate.delay` be respected? It is a delay before the animation starts playing, so it's confusing either way.
2. If the answer to (1) is yes, then how would you start it playing? A method like `Animate.play`? Thus far all playback control is delegated to the `AnimationController`, so this would be a possibly confusing departure.
3. Because `onInit` runs when the animation plays, you wouldn't have a mechanism to access to controller if there was a delay.
4. Even with `autoPlay` you still need to save off the `Animate` or `AnimationController` in order to play it later, so I'm not certain this is providing a ton of value.
5. `AnimateList` uses `Animate.delay` for `interval`, which could cause confusion`autoPlay` if didn't respect `delay`

Something to consider, but for now I'm leaning towards not adding this. It's an advanced convenience feature, that may add more confusion than value.

**Comment by esDotDev** [comment]
Thoughts:
1. I don't think it make sense to use a delay with autoPlay = false. The entire intent of autoPlay=false, is "I will handle the play call myself, at a later pt in time". In such a situation, it's trivial to add any delay I need. ie:
```
onPressed: () => Future.delayed(100.ms).then(() => _fade.play())
```
2. Non issues if we ignore delay when autoPlay:false
3.  ''
4. True, its not a LOC saving, its more readable as you can parse it down the left-hand side of the widget config.
  - Also you might pass in your own controller? In which case, its quite a bit nicer to not have to use the closure.
5. This seems like an implementation detail of AnimateList to work around... which itself is only syntactic sugar. I wouldn't let it block improvements to the core API.

**Comment by bdlindsay** [comment]
Just started trying this package out today - I immediately tried to look for an bool autoplay sort of flag. Is there some reason onInit doesn't exist on the Animate(child:) widget as well as the .animate() call? Just wanted to mention that I would've found this feature useful - thanks!

**Comment by gskinner** [comment]
Thanks for the feedback @bdlindsay. 

To answer your question: Yes, onInit was renamed onPlay because that's actually when it was firing. I had added an onInit as well, but never committed it because it gets a bit messy with `delay` and controller initialization.

Thinking out loud a bit here, but one way to think about `autoPlay=false` is simply as an infinite delay. In fact, one of the things I don't like about `autoPlay` is that it would require that `delay` is ignored if it's set to `false`. We could:

- simply document that users could use `delay: Duration(days: 0xFFFFFFFF)` to prevent autoplay
- we could add an extension `Duration.eon` (or just `eon` keyword?) to make that even easier
- we could even recognize that value and just not run the delay

Just a thought.

**Comment by esDotDev** [comment]
Alternate option, just document that delay is ignored if autoPlay = false, it doesn't really make semantic sense anyways, if there is no autoplay, what is delayed.... Once the user has taken responsibility to call `play` it's trivial for them to add their own delay.

**Comment by gskinner** [comment]
Totally, I'm always just a little hesitant to add params that are basically just a shortcut to another param. Not a huge issue though.

**Comment by bdlindsay** [comment]
Is there something wrong with just respecting the delay when the user has taken the responsibility to call play themselves if if they could add their own at that time? 
If going the direction of passing some sort of infinite duration to delay, I wonder if we could make the intended purpose of the extension more explicit. Something like `Duration.untilPlayed` or the like.

I think I would lean towards an autoPlay parameter where the specified delay could still be respected. It is easiest to discover and understand for new users of the package where an infinite duration delay seems like a bit of a hack. Delay still serves a purpose in the API to let someone specify how they expect that animation to be used, and the delay would not have to be remembered if there happened to be multiple call sites for playing the animation. It would be more confusing to know delay is ignored than to know you could remove always the delay yourself if you want to control it else where.

**Comment by esDotDev** [comment]
In order to do that, I believe a custom controller would have to be introduced. It could then define a `play` method, and internally apply the delay when user calls play.

When I made a similar lib tween lib, this use case specifically was what tipped the scales for me to implement a custom controller, rather than just passing along the raw AnimationController as animate does now.

**Comment by bdlindsay** [comment]
```dart
Animate(
  onPlay: (controller) {
    _controller ??= controller;
    if (!_isLoading) {
      controller.stop();
    }
  },
  controller: _controller,
  effects: const [RotateEffect(duration: Duration(milliseconds: 600))],
  child: const Icon(Icons.refresh, color: activeColor),
),
```

I'm just looking to spin a refresh icon as a network call is finishing, so I don't want the animation to play on initial load, just when the refresh button is pressed. In the onPress handler of the button this is contained in, I'm calling `_controller?.repeat();` to start the animation and `_controller?.stop()` when the call completes. 

Right now, I have the work-around of using a state variable to essentially just immediately stop the animation on the first page build.

**Comment by esDotDev** [comment]
> The `Animate.delay` is not part of the animation, it is a delay before the animation begins playing (ie. `AnimationController.forward()` is called), and your ability to control the animation is via the same `AnimationController`

But it could be, if `_AnimateState` were passed instead of the raw controller, or a small wrapper for the AC were created, then you could respect the defiend delay at a later call to 'play`.

The benefit to having the lib do this, is that not checking `mounted` is a potential land-mine. Proper user code needs to do something like:
```dart
onPressed: (){
  await Future.delay(1.seconds);
  if(mounted) {  // most people will forget this important check
    _controller.forward(); 
  }
}
```

This is error prone. It would be nice if I could just do: 
```dart
Animate(delay: 1.seconds, autoPlay: false, onInit: (c) => _controller = c, ...);

// then
_controller.play(); /// uses a 1.s delay, but won't fire if widget is unmounted in the meantime
```

**Comment by dokind** [comment]
how to disable autoplay right now? Guys, I want to trigger only using controller

**Comment by clragon** [comment]
since `onInit` has been removed and we only have `onPlay` now, there does not appear to be a way to stop the animation from auto-playing anymore. This means, we can't use `flutter_animate` for animations which arent automatically played on render. this feels like an oversight.

**Comment by esDotDev** [comment]
I also agree `autoPlay` is an obvious feature to have, but for now the recommened approach is to stop the animation within the onPlay callback:
```
onPlay: (c) => _controller = c..stop();
```

**Comment by clragon** [comment]
> but for now the recommended approach is to stop the animation within the onPlay callback

how so? isn't `onPlay` called everytime the animation is played?
writing that will cause the animation to not play ever again, even if we manually pass the controller and call `animateTo`.
the _only_ way to accomplish this would therefore be with the `target` value, I believe.

**Comment by esDotDev** [comment]
No, `onPlay` is called once initially, or when the animation is re-started due to widget props changing. 

But you it won't be called if you manually call `controller.forward()` yourself.

imo, it would be more clear and useful if there was a clear `onInit` callback only ever called once, and then `onPlay` could actually be dispatched _any_ time the animation starts playing, which might be handy.

**Comment by gskinner** [comment]
Now that v4.0 is released, this issue is going to be my next focus.

Here's a quick summary of my current thoughts based on the above:

- add `autoPlay` — this will prevent the internal call to `controller.forward()` in `_play`. It would also ignore `Animate.delay` entirely. This would not block `adapter` or `target` functionality.
- add `onInit` (or possibly `onInitController`) — this would fire whenever the controller is inited. Most of the time, this would only happen once, but could happen again if you pass a new controller via `controller`, or change the overall duration of the animation.
- `onPlay` — unchanged. Fires immediately after the internal `controller.forward()` call. For clarity, this wouldn't fire if `autoPlay=false`.

With regards to `onInit`, I'd love feedback on this. I also considered whether it should *only* fire in the case where a new internal controller is created. That would mean it would almost never re-init (only in extreme edge cases), but it would also make it difficult to write external logic that needs an accurate `controller.duration`. I expect that's rare, but I'm not totally convinced it's not worth supporting.

Feedback invited.

**Comment by esDotDev** [comment]
With regards to `onInit` I would expect to be basically the equivalent to `initState`, called once and only once. But in the case that you do re-create a controller, I suppose it would make sense to call again. 

I don't understand the `controller.duration` problem tbh.

**Comment by gskinner** [comment]
I just pushed a commit that adds `autoPlay` and `onInit`. I'd love to have people try it out before I do a public release so I can be sure it's meeting everyone's expectations.

It also adds some new `assert` checks to warn you when you're using incompatible features (ex. `autoPlay=false` with `onPlay` or `delay` with `adapter`).

**Comment by rienkkk** [comment]
I have tried the new autoPlay functionality but get an error when using it with an external AnimationController. I get the following error:

`The following _CastError was thrown while dispatching notifications for ValueNotifier<double>:
Null check operator used on a null value

When the exception was thrown, this was the stack
#0      Adapter.updateValue
adapter.dart:57
#1      ValueNotifierAdapter.attach.<anonymous closure>
value_notifier_adapter.dart:18
#2      ChangeNotifier.notifyListeners
change_notifier.dart:381
#3      ValueNotifier.value=
change_notifier.dart:495
#4      _CreateSheetWrapperState.build.<anonymous closure>.<anonymous closure>
create_sheet.dart:207
#5      ListenEffect.build.<anonymous closure>
listen_effect.dart:70
#6      AnimationLocalListenersMixin.notifyListeners
listener_helpers.dart:161
#7      AnimationController._tick
animation_controller.dart:837
#8      Ticker._tick
ticker.dart:249
#9      SchedulerBinding._invokeFrameCallback
binding.dart:1289
#10     SchedulerBinding.handleBeginFrame.<anonymous closure>
binding.dart:1140
#11     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:625:13)
#12     SchedulerBinding.handleBeginFrame
binding.dart:1138
#13     SchedulerBinding._handleBeginFrame
binding.dart:1055
#14     _invoke1 (dart:ui/hooks.dart:164:13)
#15     PlatformDispatcher._beginFrame (dart:ui/platform_dispatcher.dart:316:5)
#16     _beginFrame (dart:ui/hooks.dart:101:31)
The ValueNotifier<double> sending notification was: ValueNotifier<double>#eaa2e(0.6945333480834961)`


I was using it to create a ShakeOnTap widget:
`class ShakeOnTap extends StatefulWidget {
  const ShakeOnTap({
    super.key,
    required this.child,
    this.shouldShake = true,
    this.onTap,
  });

  final Widget child;
  final bool? shouldShake;
  final VoidCallback? onTap;

  @override
  State<ShakeOnTap> createState() => _ShakeOnTapState();
}

class _ShakeOnTapState extends State<ShakeOnTap> with TickerProviderStateMixin {
  late final AnimationController controller = AnimationController(
    vsync: this,
    duration: const Duration(milliseconds: 800),
  );

  @override
  Widget build(BuildContext context) {
    return Animate(
      autoPlay: false,
      controller: controller,
      effects: const [
        ShakeEffect(
          hz: 8,
          rotation: 0,
          offset: Offset(4, 0),
          duration: Duration(milliseconds: 800),
        ),
      ],
      child: AnimatedGestureDetector(
        onTap: () {
          widget.onTap?.call();
          if (widget.shouldShake ?? true) {
            controller.forward(from: 0);
          }
        },
        child: widget.child,
      ),
    );
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }
}`

**Comment by gskinner** [comment]
That error looks like it's coming from an `Adapter`, specifically a `ValueNotifierAdapter`, but I don't see one in your code. Is it perhaps something else in your app throwing that error? If so, I'd still be interested to know if it's due to a bug in the library.

**Comment by rienkkk** [comment]
@gskinner sorry, it was indeed a mistake on my side. No issues, the new autoplay functionality works perfectly!

**Comment by rienkkk** [comment]
Hi @gskinner, it seems to be indeed an issue with another Animate widget higher up the widget tree, showing up since I updated to the latest git version. I will make a new issue.

**Comment by gskinner** [comment]
Closing as resolved via 0aecdacb5418f03fe13012bc173e78daabf670d8

**Comment by benank** [comment]
Great addition! I just started using this library and needed `autoPlay. I was happy to see that it was just added. Works like a charm!
